Add Full Support for EIP-2537 (BLS12-381 Precompiles) in the Filecoin EVM

snissn opened the pull request:
### Description

This PR introduces full support for [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537) in the Filecoin EVM, implementing a suite of precompiled contracts that perform operations over the BLS12-381 elliptic curve. These precompiles enable efficient and secure cryptographic operations needed for BLS signature schemes, pairing-based proofs, and other advanced protocols. They mirror Ethereum’s spec exactly to ensure compatibility with existing tooling and cross-chain applications.

Each precompile validates input encoding, field membership, and subgroup properties as required by the EIP. Failure on malformed inputs is deterministic and burns all gas, consistent with Ethereum behavior.

---

### New Operations

The following precompiled contracts are now available at their EIP-2537-defined addresses:

| Precompile               | Address | Description                                  |
|--------------------------|---------|----------------------------------------------|
| `BLS12_G1ADD`            | `0x0b`  | Add two G1 points                            |
| `BLS12_G1MSM`            | `0x0c`  | Multi-scalar multiplication over G1          |
| `BLS12_G2ADD`            | `0x0d`  | Add two G2 points                            |
| `BLS12_G2MSM`            | `0x0e`  | Multi-scalar multiplication over G2          |
| `BLS12_PAIRING_CHECK`    | `0x0f`  | Perform a bilinear pairing check             |
| `BLS12_MAP_FP_TO_G1`     | `0x10`  | Map an Fp field element to a G1 curve point  |
| `BLS12_MAP_FP2_TO_G2`    | `0x11`  | Map an Fp2 field element to a G2 curve point |

All operations follow the ABI, encoding rules, and semantics outlined in the EIP.

---

### Testing

The test suite ensures correctness, security, and spec compliance:

#### ✅ **Success Cases**
- G1/G2 point addition, including identity and negation
- MSM over G1/G2 with valid (point, scalar) pairs and edge cases
- Mapping from Fp/Fp2 to G1/G2 with valid canonical field encodings
- Pairing tests verifying bilinearity, identity, and non-degeneracy

#### ❌ **Failure Cases**
- Invalid input lengths
- Malformed field element encoding
- Points not on curve
- Points not in subgroup (where applicable)
- Incomplete MSM or pairing input slices

#### 🧪 **Edge Behavior**
- Infinity points handled explicitly
- Zero scalars in MSM yield valid identity outputs


---

### Implementation Notes

- Backed by `blst`, a battle-tested BLS12-381 library used across the Ethereum ecosystem
- Precompiles are integrated into the EVM interpreter via the Filecoin `System` interface

---
### Remaining TODOs

- [ ] **Solidity-based test coverage**: While the Rust test suite exhaustively covers encoding, decoding, correctness, and failure behavior, we still need to add **Solidity tests** that invoke these precompiles via contracts.
---

This PR enhances FEVM’s cryptographic capabilities, aligning Filecoin with Ethereum’s tooling and enabling secure, high-performance applications that depend on BLS12-381.

Stebalien commented on actors/evm/Cargo.toml:
```diff
@@ -33,6 +33,7 @@ hex = { workspace = true }
 hex-literal = { workspace = true }
 substrate-bn = { workspace = true }
 thiserror = { workspace = true }
+blst = "0.3.14"
```
note: I checked and it turns out that this does support wasm.

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_add.rs:
```diff
@@ -0,0 +1,256 @@
+///
+/// # Safety
+///
+/// All inputs are assumed valid due to earlier checks.
```
If this function relies on the caller making checks required for _safety_, it needs to be marked unsafe. But, from what I can tell, this comment is simply incorrect and this function _doesn't_ make any assumptions about safety (e.g., the parameters are pointers).

Am I missing something?

---

Anyways... please remove/rewrite this comment regardless.

> All inputs are assumed valid due to earlier checks.

Who's responsibility is that? What does valid mean? Where are these checks? wHaT Do YoU wAnT Me tO dOOOooooOOOoOoo!

ZenGround0 replied:
After spending some time understanding rust safety I share @Stebalien's confusion on this comment.  The input params are safe immutable rust pointers owned by this code.  And the structs are declared as `#[repr(C)]` in BLST [here](https://github.com/supranational/blst/blob/master/bindings/rust/src/bindings.rs#L479) and [here](https://github.com/supranational/blst/blob/master/bindings/rust/src/bindings.rs#L429).

ZenGround0 replied:
Continuing the discussion from the fp2_to_g2 thread we should either 
1) remove this comment asserting that any inputs to this function are ok for rust safety guarantees
2) identify the invariants that must be satisfied concerning the validity of p and p_affine in order for the unsafe block to not violate rust memory safety guarantees and then label this function unsafe.

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_msm.rs:
```diff
@@ -0,0 +1,277 @@
+    // Process each (point, scalar) pair.
+    for i in 0..k {
+        let offset = i * G1_MSM_INPUT_LENGTH;
+        let encoded_point = &input[offset..offset + PADDED_G1_LENGTH];
+        let encoded_scalar = &input[offset + PADDED_G1_LENGTH..offset + G1_MSM_INPUT_LENGTH];
```
Normally I'd recommend using the `ValueReader` but... that's specifically for precompiles that treat the input as having "infinite trailing zeros". And, apparently, these new precompiles don't do that.

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_msm.rs:
```diff
@@ -0,0 +1,277 @@
+    let mut scalars: Vec<blst_scalar> = Vec::with_capacity(k);
+
+    // Process each (point, scalar) pair.
+    for i in 0..k {
```
I'd recommend iterating with `input.chunks_exact(G1_MSM_INPUT_LENGTH)`, then splitting each chunk at `PADDED_G1_LENGTH`.

snissn replied:
make sense! much cleaner 58977f60709b7e103e44f775e551671d4614c967

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/g2_add.rs:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    // Split the input into two segments corresponding to the two G2 points.
+    let a_aff = extract_g2_input(&input[..G2_INPUT_ITEM_LENGTH], false)?;
+    let b_aff = extract_g2_input(&input[G2_INPUT_ITEM_LENGTH..], false)?;
```
note: https://doc.rust-lang.org/std/primitive.slice.html#method.split_at is your friend here (and elsewhere).

Stebalien replied:
(although I haven't tried making the change, it might be worse)

snissn replied:
👍  fc764ef0

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/g2_msm.rs:
```diff
@@ -0,0 +1,334 @@
+    let mut scalars: Vec<blst_scalar> = Vec::with_capacity(k);
+
+    // Process each (point, scalar) pair.
+    for i in 0..k {
```
ditto above, I recommend using `input.chunks`

snissn replied:
👍 fc764ef0

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/map_fp2_to_g2.rs:
```diff
@@ -0,0 +1,203 @@
+/// # Safety
+///
+/// - `fp2` is assumed to be a valid Fp2 element.
+/// - The BLST function `blst_map_to_g2` is called in an unsafe block because it relies
```
are you sure? I think it's called in an unsafe block because it's an extern and takes pointers. Does it actually rely on the elements being correct?

If so, is it possible to pass this function an invalid `blst_fp2` element?

- If so, this function should be marked unsafe.
- If not, this comment should be inside this function as a "this is safe because..." comment instead of in the function documentation (it's not a constraint on the caller).

aaravm replied:
It is possible to pass an invalid element, and I have done that in the unit test cases. 
I agree with your point, and am adding the unsafe keyword to the function

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/map_fp_to_g1.rs:
```diff
@@ -0,0 +1,183 @@
+
+/// Maps an Fp field element to a G1 point (affine form).
+///
+/// # Safety
+///
+/// - `fp` is assumed to be a valid Fp element.
+/// - The BLST function `blst_map_to_g1` is called within an unsafe block,
+///   relying on the validity of inputs.
```
ditto above.

aaravm replied:
I am adding unsafe keyword here as well

aaravm replied:
updated with better comments: https://github.com/filecoin-project/builtin-actors/commit/4727f5281dfdb40382815e376da1b4dc7f2b6300

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/pairing.rs:
```diff
@@ -0,0 +1,432 @@
+    let mut pairs = Vec::with_capacity(k);
+
+    for i in 0..k {
+        let encoded_g1_element =
+            &input[i * PAIRING_INPUT_LENGTH..i * PAIRING_INPUT_LENGTH + PADDED_G1_LENGTH];
+        let encoded_g2_element = &input[i * PAIRING_INPUT_LENGTH + PADDED_G1_LENGTH
+            ..i * PAIRING_INPUT_LENGTH + PADDED_G1_LENGTH + PADDED_G2_LENGTH];
```
definitely use `input.chunks` here.

snissn replied:
👍 fc764ef0

Stebalien commented on actors/evm/src/interpreter/precompiles/bls12_381/pairing.rs:
```diff
@@ -0,0 +1,432 @@
+
+/// Helper function to perform the pairing check
+fn pairing_check(pairs: &[(blst_p1_affine, blst_p2_affine)]) -> bool {
+    if pairs.is_empty() {
+        return true;
+    }
+
+    // Compute the miller loop for the first pair
+    let (first_g1, first_g2) = &pairs[0];
```
nit: you can write this as:

```suggestion
    let Some(((first_g1, first_g2), rest)) = pairs.split_first() else {
        return true;
    };
    
    // iter over rest without having to skip anything.
```

Honestly, the compiler will probably optimize them to the same thing, but it's easier to see that this version can't panic. But I'll leave this up to you in terms of readability (my tolerance for rust shenanigans is rather high).

snissn replied:
makes sense! ff006a13

Stebalien commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,421 @@
+}
+
+/// Encodes a G1 point in affine format into byte slice with padded elements.
+pub fn encode_g1_point(input: *const blst_p1_affine) -> Vec<u8> {
```
This should be marked unsafe.

ZenGround0 replied:
Instead of keeping the *const type and labeling as unsafe you should just use the `blst_p1_affine` type. Is there a concrete reason you moved away from this?

Stebalien commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,421 @@
+
+/// https://eips.ethereum.org/EIPS/eip-2537
+/// Encodes a single finite field element into byte slice with padding.
+pub(super) fn fp_to_bytes(out: &mut [u8], input: *const blst_fp) {
```
This should me marked unsafe.

ZenGround0 replied:
Woah why are we not matching revm and inputting a val of type &blst_fp ?

Stebalien reviewed (commented):
Basic code review. Mostly just nits around safety comments.

Stebalien commented:
Ok, so, one issue here is that this now requires clang to build. It's not a show stopper, but definitely not great.

snissn commented:
Hey @aaravm I've addressed most of the feedback from @Stebalien here in this PR to your repo https://github.com/aaravm/builtin-actors/pull/2 

A few of the issues related to safety I have left for you to resolve and confirm when you next have time.


Big picture next steps beyond the points above are: 

- [ ] Create solidity test coverage
- [ ] Resolve concerns about binary size increasing due to the c library inclusion
- [ ] Review from @rvagg who can address our EIP-2537 compliance

snissn commented:
I've added an initial Solidity test that successfully uses the BLS precompile g1_add! 951f0039aa8ede64463e18fd6ae779be257e40d4 Confirmed output matches expected doubling of the generator. More coverage to come, but this validates the call path and precompile wiring end-to-end.

BigLep commented:
@snissn : are the checklist items in https://github.com/filecoin-project/builtin-actors/pull/1669#issuecomment-2892530026 complete?  If so, let @rvagg know so he can review.

snissn commented:
@BigLep yes the check list items are complete and this is read for @rvagg to review!

ZenGround0 commented on Dockerfile:
```diff
@@ -2,6 +2,9 @@ FROM rust:1.81.0-bookworm@sha256:7b7f7ae5e49819e708369d49925360bde2af4f1962842e7
 WORKDIR /usr/src/builtin-actors
 
+# Install clang
+RUN apt-get update && apt-get install -y clang
```
I want to understand if there is anything we can do in builtin-actors build.rs to make this update to clang / CC dependency as smooth as possible.  Should we be checking this dependency before we even start building or is the BLST build error going to be ergonomic enough for those building actors bundles?

My other open question is whether we have clang installed on the github actions infra that runs actor bundle generation -- we'll want to make sure that's the case to prevent jobs from failing mysteriously.

ZenGround0 replied:
Ok the above doesn't work for a few reasons but this diff does: 

```diff
diff --git a/build.rs b/build.rs
index 7dadee3c..7a0d1939 100644
--- a/build.rs
+++ b/build.rs
@@ -69,10 +69,59 @@ fn network_name() -> String {
     }.to_owned()
 }
 
+// To support EIP 2537 we compile blst library directly into evm actor
+// This requires a C => wasm32-unknown-unknown clang
+// Note that emcc while support by blst is not an option for us as it cannot
+// target wasm32-unknown-unknown
+fn check_c_compiler_does_wasm() {
+    // If CC is not set, default to "clang", else use the value from the environment.
+    let cc = match std::env::var_os("CC") {
+        Some(val) => val,
+        None => "clang".into(),
+    };
+
+    // First check cc actually runs / is in path
+    let version_check = std::process::Command::new(&cc)
+        .arg("--version")
+        .output();
+
+    if version_check.is_err() || !version_check.as_ref().unwrap().status.success() {
+        eprintln!(
+            "error: could not run C compiler '{}'.\n\
+             Please ensure clang is installed and in your PATH, or set CC to a working clang binary.",
+            cc.to_string_lossy()
+        );
+        std::process::exit(1);
+    }
+
+    // Then check that the compiler supports wasm32 target
+    let targets_check = std::process::Command::new(&cc)
+        .arg("--print-targets")
+        .output();
+
+    let has_wasm32 = match targets_check {
+        Ok(output) if output.status.success() => {
+            let targets = String::from_utf8_lossy(&output.stdout);
+            targets.lines().any(|line| line.trim().contains("wasm32"))
+        }
+        _ => false,
+    };
+
+    if !has_wasm32 {
+        eprintln!(
+            "error: C compiler '{}' does not support wasm32-unknown-unknown target.\n\
+             Install LLVM/Clang with wasm32 support \n",
+            cc.to_string_lossy()
+        );
+        std::process::exit(1);
+    }
+}
+
 fn main() -> Result<(), Box<dyn Error>> {
     // Cargo executable location.
     let cargo = std::env::var_os("CARGO").expect("no CARGO env var");
     println!("cargo:warning=cargo: {:?}", &cargo);
+    check_c_compiler_does_wasm();
 
     let out_dir = std::env::var_os("OUT_DIR")
         .as_ref()
```

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls12_381/g2_add.rs:
```diff
@@ -0,0 +1,257 @@
+///
+/// # Safety
+///
+/// All inputs are assumed valid due to earlier checks.
```
Same issue with duplicated comment

aaravm replied:
I am adding an unsafe keyword to the function to resolve this

aaravm replied:
updated with better comments: https://github.com/filecoin-project/builtin-actors/commit/4727f5281dfdb40382815e376da1b4dc7f2b6300

aaravm commented on actors/evm/src/interpreter/precompiles/bls12_381/mod.rs:
```diff
@@ -0,0 +1,15 @@
+mod g1_add;
```
Yeah, we structured it this way because each BLS precompile function has a lot of code and extensive test cases. For example, pairing.rs alone has 300 lines of tests.

BigLep commented:
@snissn : will you be able to incorporate feedback so this can get landed?

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls12_381/g2_msm.rs:
```diff
@@ -0,0 +1,332 @@
+/// **BLS12_G2MSM Precompile**
+///
+/// Implements G2 multi-scalar multiplication (MSM) according to [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537#abi-for-g2-multiexponentiation).
+/// The input must be `160 * k` bytes, where each 160-byte segment is composed of:
```
nit: This comment should say 288 not 160

ZenGround0 commented:
@aaravm @snissn would you be able to measure how much bigger this makes the evm actor's compiled wasm code?  Since we are now building all of blst inside this wasm blob we might be talking about a very large increase (i.e. 2x).  That would probably be fine but I would like to measure it before we merge to understand what is happening there.

ZenGround0 commented:
This comment is getting buried because its on an old thread: https://github.com/filecoin-project/builtin-actors/pull/1669#discussion_r2226253598.  Once we resolve this pattern one way or another I think this is ready to go in.

aaravm commented:
> This comment is getting buried because its on an old thread: [#1669 (comment)](https://github.com/filecoin-project/builtin-actors/pull/1669#discussion_r2226253598). Once we resolve this pattern one way or another I think this is ready to go in.

The function is safe and does not require unsafe on its signature because &blst_fp2 is guaranteed safe by Rust’s type system, and any variants for the output of the blst function is anyways being tested in all the different unit testcases as mentioned in https://eips.ethereum.org/EIPS/eip-2537 , and there is no possible input that can cause error.
I will edit the comment wherever I am using unsafe keyword to clarify this

snissn commented:
> @aaravm @snissn would you be able to measure how much bigger this makes the evm actor's compiled wasm code? Since we are now building all of blst inside this wasm blob we might be talking about a very large increase (i.e. 2x). That would probably be fine but I would like to measure it before we merge to understand what is happening there.

@ZenGround0 Yes I think it raises it by about 2x but it's only ~600Kb.

I had done some analysis with Steb on possible ways to optimize the wasm output for size. There's an old wasm optimization thread here: https://github.com/filecoin-project/builtin-actors/issues/1666 and a Slack message here https://filecoinproject.slack.com/archives/C015KQQLQQ1/p1745506503004619?thread_ts=1745447750.777439&cid=C015KQQLQQ1 

Implications of this for performance, and the working hypothesis is that the performance of the resulting code is the actual key metric and the size of the binary does not really imply the performance of it. We had compared various ways to decrease the wasm binary size, but the ultimate standard is using the O3 compiler flag which beat every space saving optimization we could throw at it.

BigLep commented:
@snissn: my understanding is that the ball is in your court to incorporate any open feedback.  Please re-request review when ready.

snissn commented:
head for filecoin-project/builtin-actors is b08f0b056a9e7e91bc05e112a4fb9f8ce3ab9d00
head for PR aaravm:master is fddd3bab592c620527fea71f63b3ce15e14227a1

use `make bundle` to make bundles for each and rename them to `builtin-actors.b08f0b056a9e7e91bc05e112a4fb9f8ce3ab9d00.car` and `builtin-actors.fddd3bab592c620527fea71f63b3ce15e14227a1.car`

in the folder for `filecoin-project/ref-fvm/tools/fvm-bench` i use this sol file and this bash script in this gist and output the result below https://gist.github.com/snissn/a0a9f238b820f6c6f8a1d6883b8cb48b

 Method | builtin-actors.b08f0b.car | builtin-actors.fddd3b.car | % Change
-- | -- | -- | --
helloWorld | 2,477,530 | 2,503,744 | +1.06%
recurseTest | 47,885,806 | 48,200,379 | +0.66%
fibIter | 2,697,880 | 2,724,095 | +0.97%

ZenGround0 commented:
> @ snissn do you know which eth full nodes use BLST? 

For completeness here are the eip 2537 bls libraries used by the major execution clients in ethereum ecosystem
* geth -- gnark 
* erigon -- gnark
* reth -- blst (default) && arkworks
* besu -- gnark (java wrapper / impl?)
* nethermind -- blst !

Because of nethermind something like 40% of eth nodes have been using blst for precompiles

BigLep commented:
I re-requested review from @ZenGround0 as per [2025-08-13 conversation](https://github.com/filecoin-project/builtin-actors/pull/1669#issuecomment-3186021037), the preconditions have been met.

ZenGround0 commented:
I've run into a confusing error on wasm32 build while testing this out locally as predicted in [this thread](https://github.com/filecoin-project/builtin-actors/pull/1669#discussion_r2193362205) @snissn could you please add in the preemptive check and nice error message to build.rs as suggested?

ZenGround0 commented:
> @aaravm @snissn would you be able to measure how much bigger this makes the evm actor's compiled wasm code?

FYI it is only about a 25% increase: 

-  -rwxr-xr-x    1 zenground0  staff   691K Apr  3 20:09 fil_actor_evm.wasm
+  -rwxr-xr-x    1 zenground0  staff   851K Aug 15 10:58 fil_actor_evm.wasm

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,421 @@
+/// Note: The base field is used to define G1 and G2 elements.
+pub const FP_LENGTH: usize = 48;
+
+/// SCALAR_LENGTH_BITS specifies the number of bits needed to represent an Fr element.
```
It's not your fault since it comes from the EIP but it is really confusing that we use both Fr and Fq i.e. [here](https://github.com/filecoin-project/builtin-actors/pull/1669/files#diff-da47f52af3deab18cfd2f0207d5d17fb803367eb6b047d815170c0a740f05da3R105) Please let's be consistent and pick one

ZenGround0 replied:
Ha yeah its really not your fault this notation problem is in revm source: https://github.com/bluealloy/revm/blob/6697f70dd3165e5149a0342b29d6f23f7b082cac/crates/precompile/src/bls12_381_const.rs#L132.  I'm going to recommend just changing this to Fq everywhere since that is what most of the revm code does and is the FIPs notation

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,421 @@
+}
+
+/// Encodes a G2 point in affine format into byte slice with padded elements.
+/// G2 points have two coordinates (x,y) where each coordinate is a complex number (real,imaginary)
```
```suggestion
/// G2 points have two coordinates (x,y).  Each coordinate is in field Fp2.  Each element of Fp2 is made up of two elements of Fp.  Therefore a useful analogy is that each coordinate is made up of a complex number (real,imaginary)
```

ZenGround0 replied:
Apparently there are some good analogies to real and complex in the Fp2 field extension but I'd rather we clarify that this is not exactly true

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,422 @@
+        // Basic curve check (less expensive, sufficient for addition)
+        // Only verifies that the point is on the curve
+        // SAFETY: point is properly initialized above
+        if unsafe { !blst_p2_affine_on_curve(&point) } {
```
Note to self -- this is different factoring than revm, possibly a difference in implementation.  Will revisit during review of callers.

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,422 @@
+
+    let point = decode_g1_on_curve(x_bytes, y_bytes)?;
+
+    // Check if point is on curve (no subgroup check needed for addition)
+    if subgroup_check && unsafe { !blst_p1_affine_in_g1(&point) } {
```
This comment does not match the code. We are in fact doing a subgroup check.

ZenGround0 replied:
Also this comment is referring to addition but this is the general utils package and these functions should be usable in any context.

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,422 @@
+
+pub const G1_INPUT_LENGTH: usize = 128;
+pub const G1_ADD_INPUT_LENGTH: usize = G1_INPUT_LENGTH * 2;
+pub const G1_OUTPUT_LENGTH: usize = 128;
```
This should just be replaced with PADDED_G1_LENGTH

aaravm replied:
updated: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/bfc77fae5a92a4f6b22a747cc178f75cec06125a

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,422 @@
+    blst_scalar_from_bendian,
+};
+
+pub const G1_INPUT_LENGTH: usize = 128;
```
This should just be replaced with padded G1 length

ZenGround0 replied:
Yeah it looks like revm does it the same way I'm imagining: https://github.com/bluealloy/revm/blob/6697f70dd3165e5149a0342b29d6f23f7b082cac/crates/precompile/src/bls12_381_const.rs#L145

aaravm commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_add.rs:
```diff
@@ -0,0 +1,255 @@
+                ),
+            ),
+            (
+                // bls_g1add_g1_wrong_order+g1
```
Corrected: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/cdf1986fb37ec9138861ef650d57cd6db41f7731

aaravm commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_add.rs:
```diff
@@ -0,0 +1,255 @@
+            "Short input should return IncorrectInputSize error"
+        );
+
+        // Test case 3: Large input (extra byte at start)
```
added descriptive comments in all cases https://github.com/filecoin-project/builtin-actors/pull/1669/commits/cdf1986fb37ec9138861ef650d57cd6db41f7731

aaravm commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_add.rs:
```diff
@@ -0,0 +1,255 @@
+            matches!(res, Err(PrecompileError::InvalidInput)),
+            "Invalid top bytes should return InvalidInput error"
+        );
+    }
```
Added https://github.com/filecoin-project/builtin-actors/pull/1669/commits/cdf1986fb37ec9138861ef650d57cd6db41f7731

aaravm commented on actors/evm/src/interpreter/precompiles/bls12_381/g2_add.rs:
```diff
@@ -0,0 +1,257 @@
+                ),
+            ),
+            (
+                // bls_g2add_g2_wrong_order+g2
```
corrected: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/6434e6538a3e92de3738173ba94aa791e4c3432f

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls12_381/g2_msm.rs:
```diff
@@ -0,0 +1,329 @@
+            assert_eq!(res, *expected, "g2 msm result mismatch");
+        }
+    }
+    #[test]
```
nit: is there a reason you are splitting test vectors into two different loops in one function and then a separate function below?  It is fine in that it works but it is not clean or obvious why to do it this way.

aaravm replied:
Yeah, I did this so that equal size inputs are divided into either the above or below loop. (If loops were combined, we would get an error that the sizes don't match)

ZenGround0 replied:
Sorry this doesn't make sense to me.  I'm talking about this loop and the above: https://github.com/filecoin-project/builtin-actors/pull/1669/files#diff-63a8b498dbcae7a840b750ea1994107db3b437c6e131e3a5726d30ca4c577e4cR175-R179.  Why would you get an error if these loops were combined?  I don't see it.

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls12_381/g2_msm.rs:
```diff
@@ -0,0 +1,329 @@
+            "00000000000000000000000000000000197bfd0342bbc8bee2beced2f173e1a87be576379b343e93232d6cef98d84b1d696e5612ff283ce2cfdccb2cfb65fa0c00000000000000000000000000000000184e811f55e6f9d84d77d2f79102fd7ea7422f4759df5bf7f6331d550245e3f1bcf6a30e3b29110d85e0ca16f9f6ae7a000000000000000000000000000000000f10e1eb3c1e53d2ad9cf2d398b2dc22c5842fab0a74b174f691a7e914975da3564d835cd7d2982815b8ac57f507348f00 ...
+        );
+        let res = bls12_g2msm(&mut system, &input7, PrecompileContext::default());
+        assert!(matches!(res, Err(PrecompileError::InvalidInput)));
```
Missing `bls_g2msm_point_in_correct_subgroup_invalid_curve`

aaravm replied:
added: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/b48581cc37b58754f43e2c0987363e813e46a956

ZenGround0 reviewed (commented):
Ok I am finally done!  Great job @aaravm and @snissn.  I'll hold off on approving until you've had time to work through all the review comments.  But there is nothing significant holding this back from landing.  Thanks for bearing with my slow and meticulous review.

aaravm commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_msm.rs:
```diff
@@ -0,0 +1,274 @@
+
+        // Test: Invalid top bytes
+        let invalid_top = hex::decode(
+            "1000000000000000000000000000000017f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb\
```
corrected: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/88fc15979da9c3e61ce71c95efb153c4be54b7eb

aaravm commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_msm.rs:
```diff
@@ -0,0 +1,274 @@
+
+        // Test: Point not on curve
+        let not_on_curve = hex::decode(
+            "0000000000000000000000000000000017f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb\
```
corrected: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/88fc15979da9c3e61ce71c95efb153c4be54b7eb

aaravm commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_msm.rs:
```diff
@@ -0,0 +1,274 @@
+        let res = bls12_g1msm(&mut system, &[], ctx);
+        assert!(matches!(res, Err(PrecompileError::IncorrectInputSize)));
+
+        // Test: Short input
```
added names is comments: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/88fc15979da9c3e61ce71c95efb153c4be54b7eb

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls12_381/g2_msm.rs:
```diff
@@ -197,15 +197,6 @@ mod tests {
                 ),
             ),
-            (
-                // bls_g2msm_(2p2+inf)
```
Why are you removing this ? Please add it back?

ZenGround0 replied:
aha duplicate my b

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -230,6 +230,12 @@ pub(super) fn extract_g2_input(
 /// Accepts a safe reference to a `blst_fp`; the only unsafe is localized to
 /// the FFI call that writes the big-endian bytes.
 pub(super) fn fp_to_bytes(out: &mut [u8], input: &blst_fp) {
+    debug_assert_eq!(
```
Did you just add this to check if we ever hit this condition?  I think it would be reasonable to just keep this as an assert and fail the call if out buffer is inproperly sized. I know its a departure from revm but it seems more correct.

aaravm replied:
I just removed this assert, since these values are chacked for anyways just below the debug_assert. https://github.com/filecoin-project/builtin-actors/pull/1669/commits/931ba51e014446db53cdc2fac1c23effbc72494f

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -303,7 +309,9 @@ fn decode_g1_on_curve(
     // SAFETY: Out is a blst value.
-    if unsafe { !blst_p1_affine_on_curve(&out) } {
+    let on_curve = unsafe { blst_p1_affine_on_curve(&out) };
+    let is_inf = unsafe { blst_p1_affine_is_inf(&out) };
```
blst_p1_affine_on_curve checks this (https://github.com/supranational/blst/blob/master/src/e1.c#L115) and revm implementation just does the one call.  I don't see a good reason to introduce 2 ffi crossings unless you have a strong reason I am missing.

aaravm replied:
changed: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/931ba51e014446db53cdc2fac1c23effbc72494f

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -345,7 +353,9 @@ fn decode_g2_on_curve(
     //
     // SAFETY: Out is a blst value.
-    if unsafe { !blst_p2_affine_on_curve(&out) } {
+    let on_curve = unsafe { blst_p2_affine_on_curve(&out) };
```
Same comment as [this](https://github.com/filecoin-project/builtin-actors/pull/1669/files/7c251f2f24a769b8745f34ffc0128960a1aa3b4b..fb9b9741d15d0b85740041c6b1a365c00524c6d6#r2301630997)

aaravm replied:
changed: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/931ba51e014446db53cdc2fac1c23effbc72494f

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_add.rs:
```diff
@@ -55,16 +55,19 @@ pub(super) fn p1_add_affine(a: &blst_p1_affine, b: &blst_p1_affine) -> blst_p1_a
+/// 2. All possible input variants are covered by test vectors from EIP-2537
 ///
-/// All inputs are assumed valid due to earlier checks.
+/// The unsafe block is used purely for FFI calls to the BLST library.
```
This is still unresolved

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls12_381/g1_add.rs:
```diff
@@ -0,0 +1,268 @@
+
+/// Adds a G1 point in Jacobian coordinates and a G1 point in affine form.
+///
+/// Note: While this function contains an unsafe block for BLST operations,
+/// the function itself is safe because:
+/// 1. Input types (&blst_fp2) are guaranteed safe by Rust's type system
+/// 2. All possible input variants are covered by test vectors from EIP-2537
+///
+/// The unsafe block is used purely for FFI calls to the BLST library.
```
```suggestion
/// Note: While this function contains an unsafe block for BLST operations,
/// the function itself is safe because:
/// 1. input types are all defined by blst and `repr(C)` 
/// 2. blst behavior is assumed memory safe
/// 3. The unsafe block is used purely for FFI calls to the BLST library.
```

Please use this block throughout.  I remove reference to inputs type (which was incorrect here) so that you can copy paste everywherre without worrying about fixing up.

ZenGround0 commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,443 @@
+
+    let point = decode_g1_on_curve(x_bytes, y_bytes)?;
+
+    // Check if point is on curve (no subgroup check needed for addition)
```
```suggestion
```

I think just removing this comment is cleanest.

aaravm commented on actors/evm/src/interpreter/precompiles/bls_util.rs:
```diff
@@ -0,0 +1,421 @@
+use substrate_bn::CurveError;
+
+#[inline]
+pub fn p1_to_affine(p: &blst_p1) -> blst_p1_affine {
```
added: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/309488e5593eb8f4b66bf8e8b2a46f1befab8794

aaravm commented on actors/evm/tests/contracts/BLSPrecompile.sol:
```diff
@@ -0,0 +1,236 @@
+
+    /// @notice Asserts that G1 addition precompile at 0x0B correctly computes 2·P
+    function testG1Add() public view {
+        bytes memory input = hex"0000000000000000000000000000000017f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb0000000000000000000000000000000008b3f481e3aaa0f1a09e30ed741d8ae4fcf5e095d5d00af600db18cb2c04b3edd03cc744a2888ae40caa232946c5e7e100000000000000000000000000000000112b98340eee2777cc3c14163dea3ec97977ac3dc5c70da32e6e87578f44912e902ccef9efe28d ...
```
added: https://github.com/filecoin-project/builtin-actors/pull/1669/commits/51c871b2a7647ea85ae1405947faa6a3f32ca915

BigLep commented:
Validation in Lotus will be tracked in https://github.com/filecoin-project/lotus/issues/13285
